### PR TITLE
#7753: Created form for user to choose a city for weather. Created va…

### PR DESCRIPTION
**Pull request by aster06**
…lidation for forms.

### CODE REVIEW

File: web/modules/custom/weather/src/Form/WeatherUserForm.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\weather\Services\WeatherApi;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * {@inheritdoc}
+ */
```
Comment (KuzyaWkk) [inline web/modules/custom/weather/src/Form/WeatherUserForm.php:13]: You create a new class, according to the standards, you need to describe the logic of this class, and not use {@inheritDoc}

File: web/modules/custom/weather/src/Form/WeatherUserForm.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+class WeatherUserForm extends FormBase {
+
+  /**
+   * Constructor.
+   */
+  public function __construct(ConfigFactoryInterface $config_factory,
```
Comment (KuzyaWkk) [inline web/modules/custom/weather/src/Form/WeatherUserForm.php:19]: You need to write a little more than just "Constructor".
Example: Constructs a new WeatherUserForm

File: web/modules/custom/weather/src/Services/WeatherApi.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  /**
+   * Weather City Validation.
+   */
+  public function validateCityApi(string $cityName): bool {
+    $config = $this->configFactory->get('weather.settings');
+    $apiKey = $config->get('api_admin_key');
+    try {
+      $client = \Drupal::httpClient();
```
Comment (KuzyaWkk) [inline web/modules/custom/weather/src/Services/WeatherApi.php]: Forgot about Dependency Injection :)
It must be ->
$client = $this->httpClient->get('http_client');

**Review by KuzyaWkk — CHANGES_REQUESTED** [review]
The errors are described in the comments above

File: web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php [hunk 4]
```diff
@@ -20,9 +21,14 @@
 class WeatherBlock extends BlockBase implements ContainerFactoryPluginInterface {
 
   /**
-   * Variable for Config Factory.
+   * Constructs a ConfigFactory a Client object.
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, protected ConfigFactoryInterface $configFactory) {
+  public function __construct(array $configuration,
```
Comment (KuzyaWkk) [inline web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php:26]: Not sure if it`s centred correctly

File: web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php [hunk 5]
```diff
@@ -20,9 +21,14 @@
...
+                                                               $plugin_id,
+                                                               $plugin_definition,
+                              protected ConfigFactoryInterface $configFactory,
+                              protected ClientInterface $httpClient,
+  ) {
```
Comment (KuzyaWkk) [inline web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php:31]: Not sure if it`s centred correctly

**Review by KuzyaWkk — APPROVED** [review]

File: web/modules/custom/weather/src/Form/WeatherUserForm.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+use Drupal\weather\Services\WeatherApi;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Form for user to choose a city for weather block.
+ */
+class WeatherUserForm extends FormBase {
+
+  /**
+   * Constructs a WeatherApi object.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Form/WeatherUserForm.php:17]: WeatherApi object?

File: web/modules/custom/weather/src/Form/WeatherUserForm.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  public function __construct(ConfigFactoryInterface $config_factory,
+                                                             $typedConfigManager,
+                              protected WeatherApi $openWeatherClient) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Form/WeatherUserForm.php:21]: align parameters to the left

File: web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php [hunk 8]
```diff
@@ -18,14 +19,17 @@
   category: new TranslatableMarkup("Weather block")
 )]
 class WeatherBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
   /**
-   * Variable for Config Factory.
+   * Constructs a ConfigFactory a Client object.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php:24]: weird comment, what factory? what client object?

File: web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php [hunk 9]
```diff
@@ -58,8 +63,7 @@ public function getWeatherSetting() {
   public function build() {
     $city = 'Lutsk';
     $key = $this->getWeatherSetting();
-    $client = \Drupal::httpClient();
-    $res = $client->get('https://api.openweathermap.org/data/2.5/weather?q=' . $city . '&units=metric&appid=' . $key);
+    $res = $this->httpClient->get('https://api.openweathermap.org/data/2.5/weather?q=' . $city . '&units=metric&appid=' . $key);
     $body = (string) $res->getBody();
     $res = json_decode($body, TRUE);
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php:68]: You have the API service, move this logic to it.

File: web/modules/custom/weather/src/Services/WeatherApi.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+   * Weather City Validation.
+   */
+  public function validateCityApi(string $cityName): bool {
+    $config = $this->configFactory->get('weather.settings');
+    $apiKey = $config->get('api_admin_key');
+    try {
+      $client = $this->httpClient->get('http_client');
+      $res = $client->get('https://api.openweathermap.org/data/2.5/weather?q=' . $cityName . '&units=metric&appid=' . $apiKey);
+      $body = (string) $res->getBody();
+      json_decode($body, TRUE);
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Services/WeatherApi.php:45]: Do you need to get the body if the API returns 404 for the wrong city?
json_decode is most likely doing nothing as well.

File: web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php (hunk 5 above)
Comment (lavrenchuk-dev) [reply web/modules/custom/weather/src/Plugin/Block/WeatherBlock.php:31]: align parameters to the left

File: web/modules/custom/weather/src/Form/WeatherUserForm.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    $cityName = $form_state->getValue('user_form');
+    if (!$this->openWeatherClient->validateCityApi($cityName)) {
+      $form_state->setErrorByName('user_form', $this->t('Your city is not valid.'));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state): void {}
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather/src/Form/WeatherUserForm.php:75]: You can leave `@todo Implement submit function` next time

**Review by lavrenchuk-dev — APPROVED** [review]
Considering this is WIP functionality, please account for comments in your next PR

**Pull request merged**